### ISSUE TITLE
Installer wiped unselected iexternal HDD

### ISSUE THREAD

**Issue by no-mercy** [xcp-ng/xcp#442]
I've tried to install stable xcp-ng on small ssd (30gb), from external hdd with cd emulation. On disk selection selected 30gb disk, deselected 1 gb external drive, and proceed with installation (no any errors or warnings). After installation it appeared that installator committed all partition changes to external drive instead of internal.
When i tired to install it without external drive enabled, it appeared that 30gb is not enough for installation.
As a result, all data from my external hdd was wiped.

**Comment by olivierlambert** [comment]
Are you sure you didn't select your external HDD as target for the install the first time?

**Comment by no-mercy** [comment]
> Are you sure you didn't select your external HDD as target for the install the first time?

99.9% sure

**Comment by no-mercy** [comment]
Let me try to reproduce it, i'll respond a bit later.

**Comment by no-mercy** [comment]
Yup, reproduced, it partitioned my external drive again, while it was not not selected.

But on very last step of installation wizard, right before install starts, it tells you that all data will be destroyed on sda(internal, selected drive) and sdb(external drive).

**Comment by stormi** [comment]
Can you retrieve the contents of `/var/log/installer` and share it?

**Comment by nagilum99** [comment]
While the installer falsly seems to escalate to your external drive: It's true, you need roughly 60 GB for the base system. (~30 for actual OS and the same as backup for upgrades, as it just mirrors the active partitions to the backup ones)

**Comment by no-mercy** [comment]
> Can you retrieve the contents of `/var/log/installer` and share it?

[install-log.zip](https://github.com/xcp-ng/xcp/files/5387562/install-log.zip)

**Comment by no-mercy** [comment]
please don't look on other errors in log, seems like i re-downloaded iso with errors
also, to give more input :)
![IMG_3416](https://user-images.githubusercontent.com/5758026/96181424-a378f600-0f3c-11eb-9fbc-fcff2deec387.jpeg)

**Comment by nagilum99** [comment]
The dialog looks like expected. (and as said: It should break instead of just using sdb)
I wasn't correct with the space needed, it's a bit less, nontheless: More than 29. For more Info:
https://docs.citrix.com/en-us/citrix-hypervisor/install/partition-layout.html (which is the upstream...)

**Comment by stormi** [comment]
> > Can you retrieve the contents of `/var/log/installer` and share it?
> 
> BTW, it is `/tmp/install-log`, also please don't look on other errors, seems like i've re-downloaded inconsistent iso to reproduce error
> 
> [install-log.zip](https://github.com/xcp-ng/xcp/files/5387562/install-log.zip)

`/tmp/install-log` during the installation (but you only get the log, not the extra information files), and `/var/log/installer` on the installed system, a directory containing various files among which `install-log`.

**Comment by stormi** [comment]
> The dialog looks like expected. (and as said: It should break instead of just using sdb)
> I wasn't correct with the space needed, it's a bit less, nontheless: More than 29. For more Info:
> https://docs.citrix.com/en-us/citrix-hypervisor/install/partition-layout.html (which is the upstream...)

Also documented in our official docs: https://xcp-ng.org/docs/requirements.html#disk-space

**Comment by olivierlambert** [comment]
So the issue is that the installer isn't failing but trying the next disk automatically and install there without a warning, right?

**Comment by no-mercy** [comment]
> So the issue is that the installer isn't failing but trying the next disk automatically and install there without a warning, right?

correct

**Comment by no-mercy** [comment]
> > > Can you retrieve the contents of `/var/log/installer` and share it?
> > 
> > 
> > BTW, it is `/tmp/install-log`, also please don't look on other errors, seems like i've re-downloaded inconsistent iso to reproduce error
> > [install-log.zip](https://github.com/xcp-ng/xcp/files/5387562/install-log.zip)
> 
> `/tmp/install-log` during the installation (but you only get the log, not the extra information files), and `/var/log/installer` on the installed system, a directory containing various files among which `install-log`.

i'll try to gather them

**Comment by no-mercy** [comment]
> The dialog looks like expected. (and as said: It should break instead of just using sdb)
> I wasn't correct with the space needed, it's a bit less, nontheless: More than 29. For more Info:
> https://docs.citrix.com/en-us/citrix-hypervisor/install/partition-layout.html (which is the upstream...)

I believe it should not allow to select drives that are too small.
Currently if you have too small drive ONLY, it will warn that you can't install, and would not proceed to drive select dialog.
But when you have another, big enough drive, it will pass there, and give ability to select small drive. And after all, even if you have selected small drive, it will use big one.

**Comment by stormi** [comment]
Yes, we all agree on that. Does this happen even if the external drive is not used for CD emulation?

**Comment by stormi** [comment]
And if you can do more tests, it would be good to try also with Citrix Hypervisor 8.2's installation ISO (free to download, you just need to create a free Citrix account). I believe this bug is not specific to XCP-ng and is also present in Citrix Hypervisor, but being sure about that would allow us to open a bug report for their teams.

**Comment by olivierlambert** [comment]
Yes, so we can work with them to have an upstream fix :)

**Comment by no-mercy** [comment]
I've reproduced bug with boot from flash, it installed to external drive again.
I've fetched logs. Gonna try citrix when i have time
![IMG_3426](https://user-images.githubusercontent.com/5758026/96272954-72013880-0fd7-11eb-9364-d57b0ec9903d.jpeg)
[installer.tar.gz](https://github.com/xcp-ng/xcp/files/5392390/installer.tar.gz)

**Comment by stormi** [comment]
Thanks!

**Comment by no-mercy** [comment]
Okay, it is reproducing on citrix 8.2.0 as well.

Citrix screenshot:

![IMG_3429](https://user-images.githubusercontent.com/5758026/96280041-2606c180-0fe0-11eb-8fcf-f960b65a1aaa.jpeg)

Citrix logs:

[installer-citrix.tar.gz](https://github.com/xcp-ng/xcp/files/5392753/installer-citrix.tar.gz)

After install external drive partitioning:

![Screenshot_20201016_184412](https://user-images.githubusercontent.com/5758026/96280121-48004400-0fe0-11eb-8fc8-6737ddcd25dd.png)

**Comment by stormi** [comment]
I have reported it to Citrix: https://bugs.xenserver.org/browse/XSO-990

**Comment by stormi** [comment]
Could you post a screenshot of the disk selection step for the system? All your screenshots are from the Virtual Machine Storage step.

**Comment by stormi** [comment]
Ok, I think I see what happened:
* The installer found only one disk suitable for installation, so it did not display the disk selection dialog
* Then it asked you to confirm the deletion of data on the external hard disk (that had been selected as installation target)
* Then the Local Storage Repository disk selection dialog made you think you were selecting the installation target, when it was actually already defined

So, in the end, it's not a bug, and the installer did tell you that it was going to install on the external hard disk. However the UI is probably not clear enough and led to confusion and to data loss, so this could be improved.

**Comment by stormi** [comment]
Here are the improvement suggestions I made to Citrix:
* When there are several disks but only one is suitable for installation, don't skip the target selection screen entirely: instead, replace it with a notice telling the user that only one suitable disk has been found and which and let them either confirm or give up.
* Reword the Local SR selection screen in a way that makes it obvious even to new users, not familiar with the concept of SR and VM storage, what they are doing.

**Comment by nagilum99** [comment]
> So, in the end, it's not a bug, and the installer did tell you that it was going to install on the external hard disk

As I can't reproduce it rn, question is: Did it tell you _where_ it's going to install? Wipe alone is a very thin info. If could stay by select what's there if there is only one fitting target, but the message should display _what_ it's wiping (/dev/sda...)

**Comment by stormi** [comment]
It never tells where it's going to install, but it does tell you what disks it's going to wipe data from (it does not say if it's for system or for local SRs). However it does it only after all the network configuration steps.

It's really easy to reproduce in a VM: create a VM that has two disks, one 30 GB disk and one 60 GB disk, then boot the installation ISO.
